extensions/yaml_support/labgraph_monitor/generate_lg_monitor/generate_lg_monitor.py

```diff
@@ -0,0 +1,253 @@
+    """
+    serialized_graph: SerializedGraph = {
+        "name": name,
+        "nodes": {}
+    }
+
+    for node in nodes:
+        if node.grouping not in serialized_graph["nodes"]:
+            serialized_graph["nodes"][node.grouping] = {
+                "upstreams": {}
+            }
+
+        if node.upstream_node:
+
+            if node.upstream_node.grouping not in \
+             serialized_graph["nodes"][node.grouping]["upstreams"]:
+                (serialized_graph["nodes"]
+                    [node.grouping]
+                    ["upstreams"]
+                    [node.upstream_node.grouping]) = [
+                        node.upstream_message.serialize()
+                    ]
+            else:
+                (serialized_graph["nodes"]
+                    [node.grouping]
+                    ["upstreams"]
+                    [node.upstream_node.grouping]).append(
+                        node.upstream_message.serialize()
+                    )
+
+    return serialized_graph
+
+def sub_pub_grouping_map(graph: lg.Graph) -> Dict[str, str]:
+    """
+    A function that matches subscribers with their publishers
+    
+    @params:
+        graph: An instance of the computational graph
+
+    @return: A dictionary where the key is a publisher grouping
+            and the value is a dictionary of sets of topic paths subscribed and their groupings
+    """
+    sub_pub_grouping_map: Dict[str, str] = {}
+    for stream in graph.__streams__.values():
+        difference = set(stream.topic_paths).difference(LabgraphMonitorNode.in_edges)
+        if difference:
+            upstream_edge = max(difference, key=len)
+            for edge in stream.topic_paths:
+                if edge != upstream_edge:
+                    # convert SERIALIZER/SERIALIZER_INPUT_1 to its grouping Serializer
+                    edge_path = "/".join(edge.split("/")[:-1])
+                    edge_grouping = type(graph.__descendants__[edge_path]).__name__
+                    
+                    # convert SERIALIZER/SERIALIZER_INPUT_1 to its topic SERIALIZER_INPUT_1
+                    topic_path = edge.split("/")[-1]
+
+                    # convert NOISE_GENERATOR/NOISE_GENERATOR_OUTPUT to its grouping NoiseGenerator
+                    group_path = "/".join(upstream_edge.split("/")[:-1])
+                    grouping = type(graph.__descendants__[group_path]).__name__
+
+                    if grouping in sub_pub_grouping_map:
+                        sub_pub_grouping_map[grouping]["topics"].add(topic_path)
+                        sub_pub_grouping_map[grouping]["subscribers"].add(edge_grouping)
+                    else:
+                        sub_pub_grouping_map[grouping] = {
+                            "topics": {topic_path},
+                            "subscribers": {edge_grouping},
+                        }
+                    
+    return sub_pub_grouping_map
+
+def generate_graph_topology(graph: lg.Graph) -> SerializedGraph:
```

Comment:
```suggestion
def generate_graph_topology(graph: lg.Graph) -> Tuple[SerializedGraph, Dict[str, str]]:
```

Reply:
Done, but I changed it to `None` as with your previous suggestion to abstract things we don't need to return anything

Reply:
The function still needs to return SerializedGraph in case the user would like to stream the topology using **run_topology()**, if you would like to change the return type to None, **run_topology()** should be move inside **generate_graph_topology()**

Reply:
You're right. But I figured it will be easier to to have two functions instead: **generate_graph_topology() -> SerializedGraph** and **set_graph_topology() -> None**. 

The first one returns the serialized version of the graph and should be used with **run_topology()**:

```python
from extensions.graphviz_support.graphviz_support.tests.demo_graph.demo import Demo

from extensions.yaml_support.labgraph_monitor.generate_lg_monitor.generate_lg_monitor import (
    generate_graph_topology,
)

from extensions.yaml_support.labgraph_monitor.server.lg_monitor_server import (
    run_topology,
)

# Initialize a Demo graph
graph = Demo()

# Serialize its topology
topology = generate_graph_topology(graph)

# Run the WebSockets API to send the topology to Front-End
run_topology(topology)
```

The second one sets graph topology in the graph:

```python
from extensions.yaml_support.labgraph_monitor.generate_lg_monitor.generate_lg_monitor import (
    set_graph_topology,
)

graph = Demo()
set_graph_topology(graph=graph)

runner = lg.ParallelRunner(graph=graph)
runner.run()
```

Please let me know if you see a problem with this approach. I can always update it to something else

Reply:
Having too many methods might be confusing if someone is just interested in streaming an instance of the graph. Maybe having one generic method([facade method](https://www.geeksforgeeks.org/facade-method-python-design-patterns/#:~:text=Facade%20Method%20is%20a%20Structural,consists%20of%20several%20sub%2Dsystems.)) may facilitate the usage. Besides that, the above solution looks good.

Reply:
just set it up. do you think you could take a look?